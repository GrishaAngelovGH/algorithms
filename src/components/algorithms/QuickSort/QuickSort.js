import { useState } from 'react'

import ToggleButton from 'react-bootstrap/ToggleButton'

import Legend from './Legend'
import InputRange from './InputRange'
import WaveBackground from '../../WaveBackground'

/*
    Note: 
    After swapping of two elements, swapped values are pushed to their corresponding array i.e. 'swappedElements'.
    They will be colored in yellow on each step of visualization.
    Then new subarray with all values (including swapped) is pushed to 'elements'.
    In that way we have the following map:

    [[v1, v2, v3]] i.e. 'elements' <---> [[v1, v3]] i.e. 'swappedElements'
    [[v1, v2, v3],[v4, v5, v6]] i.e. 'elements' <---> [[v1, v3], [v5, v6]] i.e. 'swappedElements'
*/

const QuickSort = () => {
    const [step, setStep] = useState(0)
    const [checked, setChecked] = useState(false)

    const array = [5, 1, 3, 2, 4]
    const steps = 6

    // used only for visualization purpose
    let elements = [[...array]]
    let swappedElements = []

    const appendToElements = (elements, arr) => [...elements, [...arr]]

    const swap = (arr, i, j) => {
        [arr[i], arr[j]] = [arr[j], arr[i]]

        // add values of swapped elements, which will be colored in yellow
        swappedElements = [...swappedElements, [arr[i], arr[j]]]
    }

    const partition = (arr, left, right, pivot) => {
        // Move bounds inward until they meet
        while (left <= right) {
            while (arr[left] < pivot) {
                left++
            }

            while ((right >= left) && (arr[right] > pivot)) {
                right--
            }

            // Swap out-of-place values
            if (right > left) {
                swap(arr, left, right)

                elements = appendToElements(elements, arr)
            }
        }

        // Return first position in right partition
        return left
    }

    const quickSort = (arr, i, j) => {
        const pivotIndex = Math.floor((i + j) / 2)

        swap(arr, pivotIndex, j) // Move pivot to the end

        elements = appendToElements(elements, arr)

        // k will be the first position in the right subarray
        const k = partition(arr, i, j - 1, arr[j])

        swap(arr, k, j) // Put pivot in place

        elements = appendToElements(elements, arr)

        if ((k - i) > 1) { quickSort(arr, i, k - 1) }  // Sort left partition
        if ((j - k) > 1) { quickSort(arr, k + 1, j) }  // Sort right partition
    }

    const handleSlideChange = value => {
        setStep(value)
    }

    const handleToggle = ({ currentTarget }) => {
        currentTarget.checked ? setStep(steps) : setStep(0)

        setChecked(currentTarget.checked)
    }

    quickSort(array, 0, array.length - 1)

    return (
        <WaveBackground>
            <h1 className='text-center text-white'>Quick Sort</h1>

            {
                Object.values(elements).map((element, i) => {

                    // last array is for the final result
                    if (i === Object.values(elements).length - 1) {
                        return (
                            <div
                                key={i}
                                className='row g-0 justify-content-center text-center p-3 text-white fw-bold'
                            >
                                <h3 className='text-white'>Final Result</h3>

                                {
                                    Object.values(element).map((v, j) => (
                                        <div key={j} className='col-md-1 mx-1 bg-success rounded'>
                                            {v}
                                        </div>
                                    ))
                                }
                            </div>
                        )
                    }

                    return (
                        <div
                            key={i}
                            className='row g-0 justify-content-center text-center p-1 text-white fw-bold'
                        >
                            {
                                Object.values(element).map((v, k) => {
                                    // skip coloring of first row of elements
                                    if (i > 0) {
                                        const isSwapped = swappedElements[i - 1].includes(v)

                                        // color all elements in yellow if toggle button is clicked
                                        // otherwise color elements in yellow on each step
                                        const bgClass = !checked ?
                                            (isSwapped && step === i ? 'bg-warning' : 'bg-success') :
                                            (isSwapped ? 'bg-warning' : 'bg-success')

                                        return (
                                            <div key={k} className={`col-md-1 mx-1 ${bgClass} rounded`}>
                                                {v}
                                            </div>
                                        )
                                    }

                                    return (
                                        <div key={k} className='col-md-1 mx-1 bg-success rounded'>
                                            {v}
                                        </div>
                                    )
                                })
                            }
                        </div>
                    )
                })
            }

            <div className='row g-0 justify-content-center'>
                <div className='col-md-6'>
                    <Legend
                        items={[
                            { color: '#198754', title: 'Regular element' },
                            { color: '#ffc107', title: 'Swapped element' }
                        ]}
                    />
                </div>
            </div>

            <div className='row g-0 justify-content-center'>
                <div className='col-md-6'>
                    <h5 className='text-center text-white'>
                        Slide to see swapped elements on each step
                    </h5>

                    <ToggleButton
                        className='mb-4'
                        id='toggle-check'
                        type='checkbox'
                        variant='outline-success'
                        checked={checked}
                        onChange={handleToggle}
                    >
                        <span className='mx-1'>Show All</span>
                    </ToggleButton>

                    <InputRange
                        disabled={checked}
                        maxValue={steps}
                        value={step}
                        formatLabel={v => `${v} of ${steps}`}
                        onChange={handleSlideChange}
                    />
                </div>
            </div>
        </WaveBackground>
    )
}

export default QuickSort