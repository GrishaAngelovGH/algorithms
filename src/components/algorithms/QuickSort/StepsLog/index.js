import StepsLog from './StepsLog'

export default StepsLog