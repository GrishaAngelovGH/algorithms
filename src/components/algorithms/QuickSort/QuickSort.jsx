import { useState } from 'react'

import Button from 'react-bootstrap/Button'
import ToggleButton from 'react-bootstrap/ToggleButton'

import Legend from 'components/Legend'
import StepsLog from 'components/StepsLog'
import Slider from 'components/Slider'
import WaveBackground from 'components/WaveBackground'

import AlgorithmSteps from './AlgorithmSteps'

/*
		Note: 
		After swapping of two elements, swapped values are pushed to their corresponding array i.e. 'swappedElements'.
		They will be colored in yellow on each step of visualization.
		Then new subarray with all values (including swapped) is pushed to 'elements'.
		In that way we have the following map:

		[[v1, v2, v3]] i.e. 'elements' <---> [[v1, v3]] i.e. 'swappedElements'
		[[v1, v2, v3],[v4, v5, v6]] i.e. 'elements' <---> [[v1, v3], [v5, v6]] i.e. 'swappedElements'
*/

const QuickSort = () => {
	const [step, setStep] = useState(0)
	const [showAllSteps, setShowAllSteps] = useState(false)
	const [showLogs, setShowLogs] = useState(false)

	const array = [5, 1, 3, 2, 4]
	const steps = 6

	// used only for visualization purpose
	let elements = [[...array]]
	let swappedElements = []

	const appendToElements = (elements, arr) => [...elements, [...arr]]

	const swap = (arr, i, j) => {
		[arr[i], arr[j]] = [arr[j], arr[i]]

		// add values of swapped elements, which will be colored in yellow
		swappedElements = [...swappedElements, [arr[i], arr[j]]]
	}

	const partition = (arr, left, right, pivot) => {
		// Move bounds inward until they meet
		while (left <= right) {
			while (arr[left] < pivot) {
				left++
			}

			while ((right >= left) && (arr[right] > pivot)) {
				right--
			}

			// Swap out-of-place values
			if (right > left) {
				swap(arr, left, right)

				elements = appendToElements(elements, arr)
			}
		}

		// Return first position in right partition
		return left
	}

	const quickSort = (arr, i, j) => {
		const pivotIndex = Math.floor((i + j) / 2)

		swap(arr, pivotIndex, j) // Move pivot to the end

		elements = appendToElements(elements, arr)

		// k will be the first position in the right subarray
		const k = partition(arr, i, j - 1, arr[j])

		swap(arr, k, j) // Put pivot in place

		elements = appendToElements(elements, arr)

		if ((k - i) > 1) { quickSort(arr, i, k - 1) }  // Sort left partition
		if ((j - k) > 1) { quickSort(arr, k + 1, j) }  // Sort right partition
	}

	const handleSlideChange = value => {
		setStep(value)
	}

	const handleToggle = ({ target: { checked } }) => {
		setStep(checked ? steps : 0)

		setShowAllSteps(checked)
	}

	const handleLogsOpen = () => {
		setShowLogs(true)
	}

	const handleLogsClose = () => {
		setShowLogs(false)
	}

	quickSort(array, 0, array.length - 1)

	return (
		<WaveBackground>
			<h1 className='text-center text-white'>Quick Sort</h1>

			{
				elements.map((row, i) => {

					// the last array contains the final result
					if (i === elements.length - 1) {
						return (
							<div
								key={i}
								className='row g-0 justify-content-center text-center p-3 text-white fw-bold'
							>
								<h3 className='text-white'>Final Result</h3>

								{
									row.map((v, j) => (
										<div key={j} className='col-md-1 mx-1 bg-success rounded'>
											{v}
										</div>
									))
								}
							</div>
						)
					}

					return (
						<div
							key={i}
							className='row g-0 justify-content-center text-center p-1 text-white fw-bold'
						>
							{
								row.map((v, k) => {
									// skip coloring of first row of elements
									if (i > 0) {
										const isSwapped = swappedElements[i - 1].includes(v)

										// color all elements in yellow if toggle button is clicked
										// otherwise color elements in yellow on each step
										const bgClass = !showAllSteps ?
											(isSwapped && step === i ? 'bg-warning' : 'bg-success') :
											(isSwapped ? 'bg-warning' : 'bg-success')

										return (
											<div key={k} className={`col-md-1 mx-1 ${bgClass} rounded`}>
												{v}
											</div>
										)
									}

									return (
										<div key={k} className='col-md-1 mx-1 bg-success rounded'>
											{v}
										</div>
									)
								})
							}
						</div>
					)
				})
			}

			<div className='row g-0 justify-content-center align-items-center text-center'>
				<div className='col-md-2'>
					<Legend
						items={[
							{ color: '#198754', title: 'Regular element' },
							{ color: '#ffc107', title: 'Swapped element' }
						]}
					/>
				</div>
				<div className='col-md-2'>
					<ToggleButton
						className='w-100'
						id='toggle-check'
						type='checkbox'
						variant='outline-primary'
						checked={showAllSteps}
						onChange={handleToggle}
					>
						<span className='mx-1'>Show All</span>
					</ToggleButton>

					<Button
						className='w-100 mt-2'
						variant='outline-primary'
						onClick={handleLogsOpen}
					>
						Algorithm Logs
					</Button>
				</div>
			</div>

			<div className='row g-0 justify-content-center mt-3'>
				<div className='col-md-6'>
					<Slider
						max={steps}
						value={step}
						disabled={showAllSteps}
						onChange={handleSlideChange}
					/>

					<h5 className='text-center text-success fw-bold mt-2'>
						Slide to see swapped elements on each step
					</h5>

					{
						showLogs && (
							<StepsLog
								title={'Quick Sort Steps'}
								onClose={handleLogsClose}
							>
								<AlgorithmSteps />
							</StepsLog>
						)
					}
				</div>
			</div>
		</WaveBackground>
	)
}

export default QuickSort