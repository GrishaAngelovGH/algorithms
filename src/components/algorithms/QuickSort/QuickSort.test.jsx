import { render } from '@testing-library/react'

import QuickSort from './QuickSort'

test('should render QuickSort component', () => {
	const view = render(<QuickSort />)

	expect(view).toMatchSnapshot()
})