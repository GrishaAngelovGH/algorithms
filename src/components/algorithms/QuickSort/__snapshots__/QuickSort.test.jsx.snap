// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render QuickSort component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="waveWrapper waveAnimation"
      >
        <div
          class="waveWrapperInner bgTop"
        >
          <div
            class="wave waveTop"
          />
        </div>
        <div
          class="waveWrapperInner bgMiddle"
        >
          <div
            class="wave waveMiddle"
          />
        </div>
        <div
          class="waveWrapperInner bgBottom"
        >
          <div
            class="wave waveBottom"
          />
        </div>
      </div>
      <div
        class="wave-content"
      >
        <h1
          class="text-center text-white"
        >
          Quick Sort
        </h1>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
        >
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
        </div>
        <div
          class="row g-0 justify-content-center text-center p-3 text-white fw-bold"
        >
          <h3
            class="text-white"
          >
            Final Result
          </h3>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            1
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            2
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            3
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            4
          </div>
          <div
            class="col-md-1 mx-1 bg-success rounded"
          >
            5
          </div>
        </div>
        <div
          class="row g-0 justify-content-center align-items-center text-center"
        >
          <div
            class="col-md-2"
          >
            <div
              class="item-legend text-white"
            >
              <div
                class="legend-title"
              >
                Legend
              </div>
              <div
                class="legend-scale"
              >
                <ul
                  class="legend-labels"
                >
                  <li
                    class="fw-bold"
                  >
                    <span
                      style="background: rgb(25, 135, 84);"
                    />
                    Regular element
                  </li>
                  <li
                    class="fw-bold"
                  >
                    <span
                      style="background: rgb(255, 193, 7);"
                    />
                    Swapped element
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            class="col-md-2"
          >
            <input
              autocomplete="off"
              class="btn-check"
              id="toggle-check"
              type="checkbox"
            />
            <label
              class="w-100 btn btn-outline-primary"
              for="toggle-check"
              tabindex="0"
            >
              <span
                class="mx-1"
              >
                Show All
              </span>
            </label>
            <button
              class="w-100 mt-2 btn btn-outline-primary"
              type="button"
            >
              Algorithm Logs
            </button>
          </div>
        </div>
        <div
          class="row g-0 justify-content-center mt-3"
        >
          <div
            class="col-md-6"
          >
            <div
              class="rc-slider rc-slider-horizontal"
            >
              <div
                class="rc-slider-rail"
                style="height: 2px;"
              />
              <div
                class="rc-slider-track"
                style="left: 0%; width: 0%; background: green;"
              />
              <div
                class="rc-slider-step"
              />
              <div
                aria-disabled="false"
                aria-orientation="horizontal"
                aria-valuemax="6"
                aria-valuemin="0"
                aria-valuenow="0"
                class="rc-slider-handle"
                role="slider"
                style="left: 0%; transform: translateX(-50%); height: 28px; width: 28px; margin-left: -14px; margin-top: -14px; background-color: green; border: 0px; opacity: 1;"
                tabindex="0"
              />
            </div>
            <h5
              class="text-center text-success fw-bold mt-2"
            >
              Slide to see swapped elements on each step
            </h5>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="waveWrapper waveAnimation"
    >
      <div
        class="waveWrapperInner bgTop"
      >
        <div
          class="wave waveTop"
        />
      </div>
      <div
        class="waveWrapperInner bgMiddle"
      >
        <div
          class="wave waveMiddle"
        />
      </div>
      <div
        class="waveWrapperInner bgBottom"
      >
        <div
          class="wave waveBottom"
        />
      </div>
    </div>
    <div
      class="wave-content"
    >
      <h1
        class="text-center text-white"
      >
        Quick Sort
      </h1>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-1 text-white fw-bold"
      >
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
      </div>
      <div
        class="row g-0 justify-content-center text-center p-3 text-white fw-bold"
      >
        <h3
          class="text-white"
        >
          Final Result
        </h3>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          1
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          2
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          3
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          4
        </div>
        <div
          class="col-md-1 mx-1 bg-success rounded"
        >
          5
        </div>
      </div>
      <div
        class="row g-0 justify-content-center align-items-center text-center"
      >
        <div
          class="col-md-2"
        >
          <div
            class="item-legend text-white"
          >
            <div
              class="legend-title"
            >
              Legend
            </div>
            <div
              class="legend-scale"
            >
              <ul
                class="legend-labels"
              >
                <li
                  class="fw-bold"
                >
                  <span
                    style="background: rgb(25, 135, 84);"
                  />
                  Regular element
                </li>
                <li
                  class="fw-bold"
                >
                  <span
                    style="background: rgb(255, 193, 7);"
                  />
                  Swapped element
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <input
            autocomplete="off"
            class="btn-check"
            id="toggle-check"
            type="checkbox"
          />
          <label
            class="w-100 btn btn-outline-primary"
            for="toggle-check"
            tabindex="0"
          >
            <span
              class="mx-1"
            >
              Show All
            </span>
          </label>
          <button
            class="w-100 mt-2 btn btn-outline-primary"
            type="button"
          >
            Algorithm Logs
          </button>
        </div>
      </div>
      <div
        class="row g-0 justify-content-center mt-3"
      >
        <div
          class="col-md-6"
        >
          <div
            class="rc-slider rc-slider-horizontal"
          >
            <div
              class="rc-slider-rail"
              style="height: 2px;"
            />
            <div
              class="rc-slider-track"
              style="left: 0%; width: 0%; background: green;"
            />
            <div
              class="rc-slider-step"
            />
            <div
              aria-disabled="false"
              aria-orientation="horizontal"
              aria-valuemax="6"
              aria-valuemin="0"
              aria-valuenow="0"
              class="rc-slider-handle"
              role="slider"
              style="left: 0%; transform: translateX(-50%); height: 28px; width: 28px; margin-left: -14px; margin-top: -14px; background-color: green; border: 0px; opacity: 1;"
              tabindex="0"
            />
          </div>
          <h5
            class="text-center text-success fw-bold mt-2"
          >
            Slide to see swapped elements on each step
          </h5>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
